File: pdf-rag-ts/src/ingestion/parser.ts
import * as fs from 'fs';
import { LlamaParseReader } from 'llamaindex';

export interface PageData {
    content: string;
    pageNumber: number;
    sourceFile: string;
    totalPages: number;
}

export class DocumentParser {
    private llamaParseApiKey: string | undefined;

    constructor(llamaParseApiKey?: string) {
        this.llamaParseApiKey = llamaParseApiKey || process.env.LLAMAPARSE_API_KEY;
    }

    async parseWithLlamaParse(filePath: string): Promise<PageData[]> {
        if (!this.llamaParseApiKey) {
            throw new Error("LlamaParse API key required for document parsing");
        }

        const reader = new LlamaParseReader({
            apiKey: this.llamaParseApiKey,
            resultType: "markdown",
            splitByPage: true,
            verbose: true,
        });

        const documents = await reader.loadData(filePath);
        if (!documents || documents.length === 0) {
            throw new Error("Failed to parse document or received empty response.");
        }

        console.log(`📄 LlamaParse returned ${documents.length} pages`);

        const pagesData: PageData[] = [];
        for (let i = 0; i < documents.length; i++) {
            const doc = documents[i];
            if (doc) {
                const pageContent = doc.text.trim();
                if (pageContent) {
                    pagesData.push({
                        content: pageContent,
                        pageNumber: i + 1,
                        sourceFile: filePath,
                        totalPages: documents.length,
                    });
                    console.log(`  📄 Page ${i + 1}: ${pageContent.length} characters`);
                }
            }
        }

        console.log(`✅ Processed ${pagesData.length} non-empty pages`);
        return pagesData;
    }

    async parseLocalMarkdown(filePath: string): Promise<PageData[]> {
        const content = fs.readFileSync(filePath, 'utf-8');
        return [{
            content: content,
            pageNumber: 1,
            sourceFile: filePath,
            totalPages: 1,
        }];
    }
}

File: pdf-rag-python/src/ingestion/parser.py
import os
from typing import List, Dict, Any

from llama_cloud_services import LlamaParse


class DocumentParser:
    def __init__(self, llamaparse_api_key: str = None):
        self.llamaparse_api_key = llamaparse_api_key or os.getenv("LLAMAPARSE_API_KEY")

    def parse_with_llamaparse(self, file_path: str) -> List[Dict[str, Any]]:
        if not self.llamaparse_api_key:
            raise ValueError("LlamaParse API key required for document parsing")

        parser = LlamaParse(
            api_key=self.llamaparse_api_key,
            num_workers=1,
            check_interval=5,
            verbose=True,
            split_by_page=True,
        )

        result = parser.parse(file_path)
        if not result:
            raise Exception("Failed to parse document or received empty response.")

        documents = result.get_markdown_documents(split_by_page=True)
        if not documents:
            raise Exception("No documents returned from LlamaParse")

        print(f"📄 LlamaParse returned {len(documents)} pages")

        pages_data = []
        for i, doc in enumerate(documents):
            page_content = doc.text.strip()
            if page_content:
                pages_data.append({
                    "content": page_content,
                    "page_number": i + 1,
                    "source_file": file_path,
                    "total_pages": len(documents),
                })
                print(f"  📄 Page {i + 1}: {len(page_content)} characters")

        print(f"✅ Processed {len(pages_data)} non-empty pages")
        return pages_data

    def parse_local_markdown(self, file_path: str) -> List[Dict[str, Any]]:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        return [{
            "content": content,
            "page_number": 1,
            "source_file": file_path,
            "total_pages": 1,
        }]
